Full jit compiled operator: cuda-gen backend

YohannDudouit opened the pull request:
This PR proposes a Cuda backend based on nvrtc, contrary to previous approaches where each operation (`BasisApply`, `QFunctionApply`, etc..) was calling a GPU kernel, this backend fuse all these into one kernel, minimising global memory read/writes. This also avoids the storage of the big E-vector, thus this backend can handle substantially larger meshes.

In order to fuse everything in one kernel I had to use a different way to write the `qFunction` compared to other cuda backends, however I can standardize all Cuda backends to use this `.qf` file format.

Early benchmarking are really good, offering substantial speed-ups over other GPU backends. Similar optimizations as the ones in #269 still need to be done though.

YohannDudouit commented:
The way the `.qf` file is written has a big impact on performance:
```c++
/// libCEED Q-function for applying a diff operator
inline __device__ void f_apply_diff(void *ctx, const CeedInt Q, const CeedScalar *ug, const CeedScalar *qd, CeedScalar *vg) {
  BuildContext *bc = (BuildContext*)ctx;
  // in[0], out[0] have shape [dim, nc=1, Q]
  for (CeedInt i = 0; i < Q; ++i) {
    const CeedScalar ug0 = ug[i+Q*0];
    const CeedScalar ug1 = ug[i+Q*1];
    vg[i+Q*0] = qd[i+Q*0]*ug0 + qd[i+Q*1]*ug1;
    vg[i+Q*1] = qd[i+Q*1]*ug0 + qd[i+Q*2]*ug1;
  }
}
```
is two time faster in 2D than:
```c++
/// libCEED Q-function for applying a diff operator
inline __device__ void f_apply_diff(void *ctx, const CeedInt Q, const CeedScalar *ug, const CeedScalar *qd, CeedScalar *vg) {
  BuildContext *bc = (BuildContext*)ctx;
  // in[0], out[0] have shape [dim, nc=1, Q]
  switch (bc->dim) {
  case 1:
    for (int i = blockIdx.x * blockDim.x + threadIdx.x;
         i < Q;
         i += blockDim.x * gridDim.x) {
      vg[i] = ug[i] * qd[i];
    }
    break;
  case 2:
    for (CeedInt i = 0; i < Q; ++i) {
      const CeedScalar ug0 = ug[i+Q*0];
      const CeedScalar ug1 = ug[i+Q*1];
      vg[i+Q*0] = qd[i+Q*0]*ug0 + qd[i+Q*1]*ug1;
      vg[i+Q*1] = qd[i+Q*1]*ug0 + qd[i+Q*2]*ug1;
    }
    break;
  case 3:
    for (CeedInt i = 0; i < Q; ++i) {
      const CeedScalar ug0 = ug[i+Q*0];
      const CeedScalar ug1 = ug[i+Q*1];
      const CeedScalar ug2 = ug[i+Q*2];
      vg[i+Q*0] = qd[i+Q*0]*ug0 + qd[i+Q*1]*ug1 + qd[i+Q*2]*ug2;
      vg[i+Q*1] = qd[i+Q*1]*ug0 + qd[i+Q*3]*ug1 + qd[i+Q*4]*ug2;
      vg[i+Q*2] = qd[i+Q*2]*ug0 + qd[i+Q*4]*ug1 + qd[i+Q*5]*ug2;
    }
    break;
  }
}
```
It seems that the `switch` case prevents the compiler from optimizing this correctly.

tcew commented:
Is this related to the cheap branching wherein all threads in a warp process all conditional options, executing nops when they are not required to do anything in a branch ?

Does the ptx show what is going on in these two different versions ?

YohannDudouit commented:
@jeremylt ,
I can test it but I have strong doubt about the capacity of the compiler to optimize anything through a double pointer.

I called it `.qf` because these files are not compiled, they are just jit compiled, so I didn't know what extension to use, but I'm not sure it should be something that gets compiled. However, I imagine I could put everything in the `.cu` file and rewrite the other kernels in terms of these functions. Even if we don't use the compiled code, it could be useful to the user to know at the compilation that the qFunction compiles, instead of discovering it at runtime (during the jit compilation). As mentioned in the description of this PR, I could also rewrite the qFunction code of the cuda backend to use these files directly.

I don't think that reusing a `qFunction` is a problem, I don't see why it would. For Occa it might be a problem because the kernels are cached, but this is not the case in my backends.

It is not possible to generate these functions from the pre-existing ones, but I can rewrite the pre-existing ones to use these functions. As mentioned in #234 I wanted to finish this backend before approving #234 , I need to test the impact on performance of your qFunctions formats to give a rigorous answer for #234 .

jeremylt commented:
I figured the double pointers might be harder to optimize, but this seemed like the right place to test and decide for all qfunctions.

At this point it's pretty close to a pure C function, with `inline __device__ ` added. If it was pure C, we could put it in a `.h` and appropriately wrap it to be a device function, and then we're closer to single source.

Yeah, I think they build the kernel once and reuse the cached kernel with insufficient memory allocated for larger problems. Just thought I'd ask to avoid the issue.

That's exactly what I was thinking about. I think we're close to getting single source for CUDA and CPU with this and #234. I suspect that the compiler on the CPU side will give better results if we can include the loop over Q in the QFunction so we can use `#pragma omp simd`, and I'm hoping that using a constant `Q=1` lets NVCC intelligently optimize on the CUDA side. @jedbrown, thoughts?

YohannDudouit commented:
@tcew I think these parts of the ptx is where the most meaningful difference between with or without `switch` happens. As we can see the data are stored in unified cache with the `switch`, whereas they are just streamed without the `switch`.
without the `switch`:
```ptx
	ld.global.f64 	%fd811, [%rd75];
	ld.global.f64 	%fd812, [%rd75+49545216];
	mul.f64 	%fd813, %fd362, %fd812;
	fma.rn.f64 	%fd814, %fd346, %fd811, %fd813;
	ld.global.f64 	%fd815, [%rd75+99090432];
	fma.rn.f64 	%fd816, %fd785, %fd815, %fd814;
	ld.global.f64 	%fd817, [%rd75+148635648];
	mul.f64 	%fd818, %fd362, %fd817;
	fma.rn.f64 	%fd819, %fd346, %fd812, %fd818;
	ld.global.f64 	%fd820, [%rd75+198180864];
	fma.rn.f64 	%fd378, %fd785, %fd820, %fd819;
	mul.f64 	%fd821, %fd362, %fd820;
	fma.rn.f64 	%fd822, %fd346, %fd815, %fd821;
	ld.global.f64 	%fd823, [%rd75+247726080];
	fma.rn.f64 	%fd379, %fd785, %fd823, %fd822;
	ld.global.f64 	%fd824, [%rd75+288];
	ld.global.f64 	%fd825, [%rd75+49545504];
	mul.f64 	%fd826, %fd363, %fd825;
	fma.rn.f64 	%fd827, %fd347, %fd824, %fd826;
	ld.global.f64 	%fd828, [%rd75+99090720];
	fma.rn.f64 	%fd829, %fd790, %fd828, %fd827;
	ld.global.f64 	%fd830, [%rd75+148635936];
	mul.f64 	%fd831, %fd363, %fd830;
	fma.rn.f64 	%fd832, %fd347, %fd825, %fd831;
	ld.global.f64 	%fd833, [%rd75+198181152];
	fma.rn.f64 	%fd380, %fd790, %fd833, %fd832;
	mul.f64 	%fd834, %fd363, %fd833;
	fma.rn.f64 	%fd835, %fd347, %fd828, %fd834;
	ld.global.f64 	%fd836, [%rd75+247726368];
	fma.rn.f64 	%fd381, %fd790, %fd836, %fd835;
	ld.global.f64 	%fd837, [%rd75+576];
	ld.global.f64 	%fd838, [%rd75+49545792];
	mul.f64 	%fd839, %fd364, %fd838;
	fma.rn.f64 	%fd840, %fd348, %fd837, %fd839;
	ld.global.f64 	%fd841, [%rd75+99091008];
	fma.rn.f64 	%fd842, %fd795, %fd841, %fd840;
	ld.global.f64 	%fd843, [%rd75+148636224];
	mul.f64 	%fd844, %fd364, %fd843;
	fma.rn.f64 	%fd845, %fd348, %fd838, %fd844;
	ld.global.f64 	%fd846, [%rd75+198181440];
	fma.rn.f64 	%fd382, %fd795, %fd846, %fd845;
	mul.f64 	%fd847, %fd364, %fd846;
	fma.rn.f64 	%fd848, %fd348, %fd841, %fd847;
	ld.global.f64 	%fd849, [%rd75+247726656];
	fma.rn.f64 	%fd383, %fd795, %fd849, %fd848;
	ld.global.f64 	%fd850, [%rd75+864];
	ld.global.f64 	%fd851, [%rd75+49546080];
	mul.f64 	%fd852, %fd365, %fd851;
	fma.rn.f64 	%fd853, %fd349, %fd850, %fd852;
	ld.global.f64 	%fd854, [%rd75+99091296];
	fma.rn.f64 	%fd855, %fd800, %fd854, %fd853;
	ld.global.f64 	%fd856, [%rd75+148636512];
	mul.f64 	%fd857, %fd365, %fd856;
	fma.rn.f64 	%fd858, %fd349, %fd851, %fd857;
	ld.global.f64 	%fd859, [%rd75+198181728];
	fma.rn.f64 	%fd384, %fd800, %fd859, %fd858;
	mul.f64 	%fd860, %fd365, %fd859;
	fma.rn.f64 	%fd861, %fd349, %fd854, %fd860;
	ld.global.f64 	%fd862, [%rd75+247726944];
	fma.rn.f64 	%fd385, %fd800, %fd862, %fd861;
	ld.global.f64 	%fd863, [%rd75+1152];
	ld.global.f64 	%fd864, [%rd75+49546368];
	mul.f64 	%fd865, %fd366, %fd864;
	fma.rn.f64 	%fd866, %fd350, %fd863, %fd865;
	ld.global.f64 	%fd867, [%rd75+99091584];
	fma.rn.f64 	%fd868, %fd805, %fd867, %fd866;
	ld.global.f64 	%fd869, [%rd75+148636800];
	mul.f64 	%fd870, %fd366, %fd869;
	fma.rn.f64 	%fd871, %fd350, %fd864, %fd870;
	ld.global.f64 	%fd872, [%rd75+198182016];
	fma.rn.f64 	%fd386, %fd805, %fd872, %fd871;
	mul.f64 	%fd873, %fd366, %fd872;
	fma.rn.f64 	%fd874, %fd350, %fd867, %fd873;
	ld.global.f64 	%fd875, [%rd75+247727232];
	fma.rn.f64 	%fd387, %fd805, %fd875, %fd874;
	ld.global.f64 	%fd876, [%rd75+1440];
	ld.global.f64 	%fd877, [%rd75+49546656];
	mul.f64 	%fd878, %fd367, %fd877;
	fma.rn.f64 	%fd879, %fd351, %fd876, %fd878;
	ld.global.f64 	%fd880, [%rd75+99091872];
	fma.rn.f64 	%fd881, %fd810, %fd880, %fd879;
	ld.global.f64 	%fd882, [%rd75+148637088];
	mul.f64 	%fd883, %fd367, %fd882;
	fma.rn.f64 	%fd884, %fd351, %fd877, %fd883;
	ld.global.f64 	%fd885, [%rd75+198182304];
	fma.rn.f64 	%fd388, %fd810, %fd885, %fd884;
	mul.f64 	%fd886, %fd367, %fd885;
	fma.rn.f64 	%fd887, %fd351, %fd880, %fd886;
	ld.global.f64 	%fd888, [%rd75+247727520];
	fma.rn.f64 	%fd389, %fd810, %fd888, %fd887;
	fma.rn.f64 	%fd889, %fd81, %fd816, 0d0000000000000000;
	fma.rn.f64 	%fd890, %fd82, %fd829, %fd889;
	fma.rn.f64 	%fd891, %fd83, %fd842, %fd890;
	fma.rn.f64 	%fd892, %fd84, %fd855, %fd891;
	fma.rn.f64 	%fd893, %fd85, %fd868, %fd892;
	fma.rn.f64 	%fd894, %fd86, %fd881, %fd893;
	fma.rn.f64 	%fd895, %fd87, %fd816, 0d0000000000000000;
	fma.rn.f64 	%fd896, %fd88, %fd829, %fd895;
	fma.rn.f64 	%fd897, %fd89, %fd842, %fd896;
	fma.rn.f64 	%fd898, %fd90, %fd855, %fd897;
	fma.rn.f64 	%fd899, %fd91, %fd868, %fd898;
	fma.rn.f64 	%fd390, %fd92, %fd881, %fd899;
	fma.rn.f64 	%fd900, %fd93, %fd816, 0d0000000000000000;
	fma.rn.f64 	%fd901, %fd94, %fd829, %fd900;
	fma.rn.f64 	%fd902, %fd95, %fd842, %fd901;
	fma.rn.f64 	%fd903, %fd96, %fd855, %fd902;
	fma.rn.f64 	%fd904, %fd97, %fd868, %fd903;
	fma.rn.f64 	%fd391, %fd98, %fd881, %fd904;
	fma.rn.f64 	%fd905, %fd99, %fd816, 0d0000000000000000;
	fma.rn.f64 	%fd906, %fd100, %fd829, %fd905;
	fma.rn.f64 	%fd907, %fd101, %fd842, %fd906;
	fma.rn.f64 	%fd908, %fd102, %fd855, %fd907;
	fma.rn.f64 	%fd909, %fd103, %fd868, %fd908;
	fma.rn.f64 	%fd392, %fd104, %fd881, %fd909;
	fma.rn.f64 	%fd910, %fd105, %fd816, 0d0000000000000000;
	fma.rn.f64 	%fd911, %fd106, %fd829, %fd910;
	fma.rn.f64 	%fd912, %fd107, %fd842, %fd911;
	fma.rn.f64 	%fd913, %fd108, %fd855, %fd912;
	fma.rn.f64 	%fd914, %fd109, %fd868, %fd913;
	fma.rn.f64 	%fd393, %fd110, %fd881, %fd914;
```
with the `switch`:
```ptx
	ld.global.f64 	%fd384, [%rd84];
	ld.global.f64 	%fd390, [%rd84+288];
	ld.global.f64 	%fd385, [%rd84+49545216];
	ld.global.f64 	%fd391, [%rd84+49545504];
	ld.global.f64 	%fd386, [%rd84+99090432];
	ld.global.f64 	%fd392, [%rd84+99090720];
	ld.global.f64 	%fd387, [%rd84+148635648];
	ld.global.f64 	%fd393, [%rd84+148635936];
	ld.global.f64 	%fd388, [%rd84+198180864];
	ld.global.f64 	%fd394, [%rd84+198181152];
	ld.global.f64 	%fd389, [%rd84+247726080];
	ld.global.f64 	%fd395, [%rd84+247726368];
	ld.global.f64 	%fd402, [%rd84+864];
	ld.global.f64 	%fd396, [%rd84+576];
	ld.global.f64 	%fd403, [%rd84+49546080];
	ld.global.f64 	%fd397, [%rd84+49545792];
	ld.global.f64 	%fd404, [%rd84+99091296];
	ld.global.f64 	%fd398, [%rd84+99091008];
	ld.global.f64 	%fd405, [%rd84+148636512];
	ld.global.f64 	%fd399, [%rd84+148636224];
	ld.global.f64 	%fd406, [%rd84+198181728];
	ld.global.f64 	%fd400, [%rd84+198181440];
	ld.global.f64 	%fd407, [%rd84+247726944];
	ld.global.f64 	%fd401, [%rd84+247726656];
	ld.global.f64 	%fd414, [%rd84+1440];
	ld.global.f64 	%fd408, [%rd84+1152];
	ld.global.f64 	%fd415, [%rd84+49546656];
	ld.global.f64 	%fd409, [%rd84+49546368];
	ld.global.f64 	%fd416, [%rd84+99091872];
	ld.global.f64 	%fd410, [%rd84+99091584];
	ld.global.f64 	%fd417, [%rd84+148637088];
	ld.global.f64 	%fd411, [%rd84+148636800];
	ld.global.f64 	%fd418, [%rd84+198182304];
	ld.global.f64 	%fd412, [%rd84+198182016];
	ld.global.f64 	%fd419, [%rd84+247727520];
	ld.global.f64 	%fd413, [%rd84+247727232];
	st.local.v2.f64 	[%rd2], {%fd384, %fd390};
	st.local.v2.f64 	[%rd2+48], {%fd385, %fd391};
	st.local.v2.f64 	[%rd2+96], {%fd386, %fd392};
	st.local.v2.f64 	[%rd2+144], {%fd387, %fd393};
	st.local.v2.f64 	[%rd2+192], {%fd388, %fd394};
	st.local.v2.f64 	[%rd2+240], {%fd389, %fd395};
	st.local.v2.f64 	[%rd2+16], {%fd396, %fd402};
	st.local.v2.f64 	[%rd2+64], {%fd397, %fd403};
	st.local.v2.f64 	[%rd2+112], {%fd398, %fd404};
	st.local.v2.f64 	[%rd2+160], {%fd399, %fd405};
	st.local.v2.f64 	[%rd2+208], {%fd400, %fd406};
	st.local.v2.f64 	[%rd2+256], {%fd401, %fd407};
	st.local.v2.f64 	[%rd2+32], {%fd408, %fd414};
	st.local.v2.f64 	[%rd2+80], {%fd409, %fd415};
	st.local.v2.f64 	[%rd2+128], {%fd410, %fd416};
	st.local.v2.f64 	[%rd2+176], {%fd411, %fd417};
	st.local.v2.f64 	[%rd2+224], {%fd412, %fd418};
	st.local.v2.f64 	[%rd2+272], {%fd413, %fd419};
```

YohannDudouit commented:
I solve the switch performance issue, this was my fault. I had left a `for` loop in the 1d case that the compiler couldn't optimize; this was preventing the compiler to optimize 2d and 3d cases. So we don't have to split the qFunctions for each dimension.

jedbrown commented on Makefile:
```diff
@@ -61,6 +67,7 @@ AFLAGS = -fsanitize=address #-fsanitize=undefined -fno-omit-frame-pointer
 
 OPT    = -O -g -march=native -ffp-contract=fast -fopenmp-simd
 CFLAGS = -std=c99 $(OPT) -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP
+CXXFLAGS = $(OPT) -Wall -Wextra -Wno-unused-parameter -fPIC 
```
Can we include `-MMD` here so that we get automatic dependencies?

YohannDudouit replied:
Should I also add the `-MP` flag?

jedbrown replied:
Yes, please.

jedbrown commented on Makefile:
```diff
@@ -201,6 +211,7 @@ info:
 	$(info FC         = $(FC))
```
Probably should mention `CXX` here since we're mentioning `CXXFLAGS`.

jedbrown commented on Makefile:
```diff
@@ -280,13 +291,16 @@ endif
+CUDA_BACKENDS = /gpu/cuda/ref /gpu/cuda/reg /gpu/cuda/shared /gpu/cuda/gen
 ifneq ($(CUDA_LIB_DIR),)
   $(libceed) : CFLAGS += -I$(CUDA_DIR)/include
+  $(libceed) : CXXFLAGS += -I$(CUDA_DIR)/include
```
The above line is `CPPFLAGS` in master and that should be sufficient for building C++.

jedbrown commented on backends/cuda-gen/ceed-cuda-gen-operator.c:
```diff
@@ -0,0 +1,199 @@
+  // }
+  // ierr = CeedFree(&impl->qvecsout); CeedChk(ierr);
+
+  // ierr = CeedFree(&impl->W); CeedChk(ierr);
```
Should this commented code be executed?

YohannDudouit replied:
Thanks, I completely forgot about the destructors.

jedbrown commented on backends/cuda-gen/ceed-cuda-gen-qfunction.c:
```diff
@@ -0,0 +1,126 @@
+  ierr = CeedSetBackendFunction(ceed, "QFunction", qf, "Destroy",
+                                CeedQFunctionDestroy_Cuda_gen); CeedChk(ierr);
+  return 0;
+}
```
Missing newline at end of file.

YohannDudouit commented:
I discussed with @tzanio if we want this to be merged before PR #288 . The conclusion was that we want to have this backend merged in master for CEED 3AM. It could be nice to separate the two merges in two commits for bug tracing. But if you believe we can be done with PR #288 before CEED 3AM, I don't have any strong opinion.
@jedbrown @jeremylt I have trouble merging master into this branch though, I get linking errors with `libceed_test.so`.

jedbrown commented:
I think we can probably finish #288 (but there may be unexpected hiccups). I think it's fine to merge this so we have a common foundation to build on. Would you be happy squashing this for the merge (for bisectability) or would you like to achieve that by `rebase -i`?
If the link errors are new (not in 'master'), can you push (here or in a temporary branch) so we can inspect a common merge result?

YohannDudouit commented:
@jedbrown I pushed the "merge" with master. I fine with squashing this merge once the error is resolved.

YohannDudouit commented:
This PR passes all the tests and is ready for review.

jeremylt commented on backends/cuda-gen/ceed-cuda-gen-qfunction.c:
```diff
@@ -0,0 +1,125 @@
+  Ceed ceed;
+  ierr = CeedQFunctionGetCeed(qf, &ceed); CeedChk(ierr);
+  return CeedError(ceed, 1, "Backend does not implement QFunctionApply");
+}
```
One suggestion - Do we want to delegate this to a different CUDA backend just in case `CeedQFunctionApply` is important to the user's code? I'm not sure it's possible, but I thought I would ask. The user can always make a second CEED to do it themselves.

YohannDudouit replied:
It's not so simple to delegate because the delegate doesn't use the same qFunction file for the moment, I can add the delegation to #288 .

YohannDudouit commented:
@jedbrown Are you still reviewing this PR, or can I squash and merge it as is?

jedbrown commented:
@YohannDudouit Thanks for the bump. Do you experience this issue when using ifort?
```
1..39
ok 1 t400-qfunction-f /cpu/self/ref/serial
ok 2 t400-qfunction-f /cpu/self/ref/serial stdout
ok 3 t400-qfunction-f /cpu/self/ref/serial stderr
ok 4 t400-qfunction-f /cpu/self/ref/blocked
ok 5 t400-qfunction-f /cpu/self/ref/blocked stdout
ok 6 t400-qfunction-f /cpu/self/ref/blocked stderr
ok 7 t400-qfunction-f /cpu/self/opt/serial
ok 8 t400-qfunction-f /cpu/self/opt/serial stdout
ok 9 t400-qfunction-f /cpu/self/opt/serial stderr
ok 10 t400-qfunction-f /cpu/self/opt/blocked
ok 11 t400-qfunction-f /cpu/self/opt/blocked stdout
ok 12 t400-qfunction-f /cpu/self/opt/blocked stderr
ok 13 t400-qfunction-f /cpu/self/ref/memcheck
ok 14 t400-qfunction-f /cpu/self/ref/memcheck stdout
ok 15 t400-qfunction-f /cpu/self/ref/memcheck stderr
ok 16 t400-qfunction-f /cpu/self/xsmm/serial
ok 17 t400-qfunction-f /cpu/self/xsmm/serial stdout
ok 18 t400-qfunction-f /cpu/self/xsmm/serial stderr
ok 19 t400-qfunction-f /cpu/self/xsmm/blocked
ok 20 t400-qfunction-f /cpu/self/xsmm/blocked stdout
ok 21 t400-qfunction-f /cpu/self/xsmm/blocked stderr
not ok 22 t400-qfunction-f /gpu/cuda/ref
ok 23 t400-qfunction-f /gpu/cuda/ref stdout
not ok 24 t400-qfunction-f /gpu/cuda/ref stderr
# +/home/jedbrown/libceed/backends/cuda/ceed-cuda-qfunction.c:115 in loadCudaFunction(): Couldn't open the Cuda file for the QFunction.
not ok 25 t400-qfunction-f /gpu/cuda/reg
ok 26 t400-qfunction-f /gpu/cuda/reg stdout
not ok 27 t400-qfunction-f /gpu/cuda/reg stderr
# +/home/jedbrown/libceed/backends/cuda/ceed-cuda-qfunction.c:115 in loadCudaFunction(): Couldn't open the Cuda file for the QFunction.
not ok 28 t400-qfunction-f /gpu/cuda/shared
ok 29 t400-qfunction-f /gpu/cuda/shared stdout
not ok 30 t400-qfunction-f /gpu/cuda/shared stderr
# +/home/jedbrown/libceed/backends/cuda/ceed-cuda-qfunction.c:115 in loadCudaFunction(): Couldn't open the Cuda file for the QFunction.
not ok 31 t400-qfunction-f /gpu/cuda/gen
ok 32 t400-qfunction-f /gpu/cuda/gen stdout
not ok 33 t400-qfunction-f /gpu/cuda/gen stderr
# +/home/jedbrown/libceed/backends/cuda-gen/ceed-cuda-gen-qfunction.c:63 in loadCudaFunction(): Couldn't open the Cuda file for the QFunction.
ok 34 t400-qfunction-f /cpu/self/tmpl
ok 35 t400-qfunction-f /cpu/self/tmpl stdout
ok 36 t400-qfunction-f /cpu/self/tmpl stderr
ok 37 t400-qfunction-f /cpu/self/tmpl/sub
ok 38 t400-qfunction-f /cpu/self/tmpl/sub stdout
ok 39 t400-qfunction-f /cpu/self/tmpl/sub stderr
Failed 8/39 subtests 
```

That looks like the path issue that we've encountered before. These standard tests all run cleanly for me with gcc-7 and 8.

Have you tried `make prove search=petsc`? All other backends (including the other GPU backends) work, but I get
```
ok 37 petsc-bps /gpu/cuda/gen
not ok 38 petsc-bps /gpu/cuda/gen stdout                                                       
# + Pointwise Error (max)              : 5.749779e-01   
```

jeremylt commented:
I got that same result with ifort on CU's machine and it was issue #71 in that case.

YohannDudouit commented:
@jedbrown Sorry, I sneaked an optimization and introduced a bug... I fixed everything in my last commit, I preferred to keep the optimization as it provides a nice speedup.
I have this error on Ray since the new petsc BPs (unrelated to this PR though):
```
ok 1 petsc-bps /cpu/self/ref/serial
not ok 2 petsc-bps /cpu/self/ref/serial stdout
# + [1564001939.776675] [ray21:36535:0]         mxm.c:196  MXM  WARN  The 'ulimit -s' on the system is set to 'unlimited'. This may have negative performance implications. Please set the stack size to the default value (10240)
# + [1564001939.780646] [ray21:36535:0]         mxm.c:196  MXM  WARN  The 'ulimit -s' on the system is set to 'unlimited'. This may have negative performance implications. Please set the stack size to the default value (10240)
not ok 3 petsc-bps /cpu/self/ref/serial stderr
# +--------------------------------------------------------------------------
# +WARNING: There are more than one active ports on host 'ray21', but the
# +default subnet GID prefix was detected on more than one of these
# +ports.  If these ports are connected to different physical IB
# +networks, this configuration will fail in Open MPI.  This version of
# +Open MPI requires that every physically separate IB subnet that is
# +used between connected MPI processes must have different subnet ID
# +values.
# +
# +Please see this FAQ entry for more details:
# +
# +http://www.open-mpi.org/faq/?category=openfabrics#ofa-default-subnet-gid
# +
# +NOTE: You can turn off this warning by setting the MCA parameter
# +btl_openib_warn_default_gid_prefix to 0.
# +--------------------------------------------------------------------------
# +[ray21:36535] mca_base_component_repository_open: unable to open mca_pml_pami: libpami.so.3: cannot open shared object file: No such file or directory (ignored)
# +[ray21:36535] mca_base_component_repository_open: unable to open mca_coll_hcoll: libhcoll.so.1: cannot open shared object file: No such file or directory (ignored)
# +[ray21:36535] mca_base_component_repository_open: unable to open mca_coll_ibm: libcollectives.so.3: cannot open shared object file: No such file or directory (ignored)
# +[ray21:36535] mca_base_component_repository_open: unable to open mca_osc_pami: libpami.so.3: cannot open shared object file: No such file or directory (ignored)
```
`make prove-all` and `make prove search=petsc` pass all tests otherwise.

jedbrown commented:
Hmm, there is no `mxm.c` in libCEED or PETSc. Do you know where it's coming from?

jeremylt commented:
@YohannDudouit, you see this on the current master as well?

YohannDudouit commented:
@jeremylt Same problem on the current master with the last version of petsc.
Since it's unrelated to this PR, maybe that I can open an issue, and we omit this issue for the review of this PR.

jedbrown commented:
That "mention" with an older PETSc is still not a PETSc-related error message. That file didn't exist for any of the versions mentioned. I'm fine with merging this now.
